src/wp-includes/block-template.php

```diff
@@ -228,7 +228,25 @@ function get_the_block_template_html() {
 	$content = $wp_embed->autoembed( $content );
 	$content = shortcode_unautop( $content );
 	$content = do_shortcode( $content );
-	$content = do_blocks( $content );
+
+	/*
+	 * Most block themes omit the `core/query` and `core/post-template` blocks in their singular content templates.
+	 * While this technically still works since singular content templates are always for only one post, it results in
+	 * the main query loop never being entered which causes bugs in core and the plugin ecosystem.
+	 *
+	 * The workaround below ensures that the loop is started even for those singular templates without the relevant
+	 * blocks. The while loop will by definition only go through a single iteration, i.e. `do_blocks()` is only called
+	 * once.
+	 */
+	if ( is_singular() && ! has_block( 'core/post-template', $content ) ) {
```

Comment:
My rationale for adding it was that a block theme author may still add the block to their singular template within a `core/query` block, which would trigger the main query loop anyway - which would still be slightly cleaner so I was thinking in that case we should respect that and use that approach to trigger the loop rather than globally doing it here. LMKWYT.

Reply:
I think the main use-case for query (and post template) within singular templates is without "inherit: true" for things like "related posts with similar categories" or "recent posts" or things like that. So I think this check is probably going to reintroduce the "error" for most of these themes.

My question now becomes: what happens if we drop the check and we end up running the query twice: one here and one with the `core/post-template` block (with inherit true)?

Reply:
> I think the main use-case for query (and post template) within singular templates is without "inherit: true" for things like "related posts with similar categories" or "recent posts" or things like that. So I think this check is probably going to reintroduce the "error" for most of these themes.

Makes sense. We should either remove this, or we would need to implement a more precise check that only triggers the loop if no `core/post-template` block wrapped by a `core/query` block with `inherit: true` is present - but that may be quite complicated to implement reasonably within this logic.

> My question now becomes: what happens if we drop the check and we end up running the query twice: one here and one with the `core/post-template` block (with inherit true)?

In that case, because of the fix in https://github.com/WordPress/gutenberg/pull/49904, it would just run through a "duplicated" loop within the main query loop, which would be fine though since it would be just for the same single post. Basically it would be something like this:
```php
// Main query loop is only run once for the single post.
while ( have_posts() ) {
	the_post();

	// Copy of main query loop from `core/post-template` is only run once for the same post.
	while ( $duplicated_query->have_posts() ) {
		the_post();

		// Display the post.
	}
}
```

So while that feels a bit odd, there wouldn't be an actual problem, since the duplicated loop would still be within the main query loop, so things would work as expected.

Based on that, I should probably remove the check for the `core/post-template` block.

Reply:
Updated in https://github.com/WordPress/wordpress-develop/pull/5104/commits/0726f9b85cd4f6b6ae616bb93994d48f5b4980ff.